File: go/Knight Probability in Chessboard.go
package main

func knightProbability(n, k, row, column int) float64 {
	// Possible knight moves
	moves := [8][2]int{{-2, -1}, {-1, -2}, {-2, 1}, {-1, 2}, {1, -2}, {2, -1}, {1, 2}, {2, 1}}

	// Initialize a 3D memoization table to store the probabilities
	dp := make([][][]float64, k+1)
	for i := range dp {
		dp[i] = make([][]float64, n)
		for j := range dp[i] {
			dp[i][j] = make([]float64, n)
		}
	}

	// Base case: probability of starting position (k=0) is 1
	dp[0][row][column] = 1.0

	// Dynamic programming to calculate probabilities
	for step := 1; step <= k; step++ {
		for r := 0; r < n; r++ {
			for c := 0; c < n; c++ {
				for _, move := range moves {
					nr, nc := r+move[0], c+move[1]
					if nr >= 0 && nr < n && nc >= 0 && nc < n {
						dp[step][r][c] += dp[step-1][nr][nc] / 8.0
					}
				}
			}
		}
	}

	// Sum all probabilities for all cells on the chessboard after k moves
	var probability float64
	for r := 0; r < n; r++ {
		for c := 0; c < n; c++ {
			probability += dp[k][r][c]
		}
	}

	return probability
}

File: bin/python3/Knight Probability in Chessboard.py
class Solution:
    def knightProbability(self, n: int, k: int, row: int, column: int) -> float:
        # Possible knight moves
        moves = [(-2, -1), (-1, -2), (-2, 1), (-1, 2), (1, -2), (2, -1), (1, 2), (2, 1)]

        # Initialize a 3D memoization table to store the probabilities
        dp = [[[0 for _ in range(n)] for _ in range(n)] for _ in range(k + 1)]

        # Base case: probability of starting position (k=0) is 1
        dp[0][row][column] = 1

        # Dynamic programming to calculate probabilities
        for step in range(1, k + 1):
            for r in range(n):
                for c in range(n):
                    for dr, dc in moves:
                        nr, nc = r + dr, c + dc
                        if 0 <= nr < n and 0 <= nc < n:
                            dp[step][r][c] += dp[step - 1][nr][nc] / 8.0

        # Sum all probabilities for all cells on the chessboard after k moves
        probability = sum(sum(dp[k][r][c] for r in range(n)) for c in range(n))

        return probability